plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -927,30 +1011,72 @@ fu_genesys_usbhub_device_setup(FuDevice *device, GError **error)
 		}
 
 		self->read_first_bank =
-		    (self->isp_model == ISP_MODEL_HUB_GL3523) && self->fw_bank_vers[0] != 0;
+		    (self->chip.model == ISP_MODEL_HUB_GL3523) && self->fw_bank_vers[0] != 0;
 		self->write_recovery_bank = address == self->fw_bank_addr[1];
 	}
 
 	/* has public key */
 	if (fu_device_has_private_flag(device, FU_GENESYS_USBHUB_FLAG_HAS_PUBLIC_KEY)) {
 		g_autofree gchar *guid = NULL;
-		if (!fu_memcpy_safe(self->public_key,
-				    sizeof(self->public_key),
-				    0, /* dst */
-				    g_bytes_get_data(blob, NULL),
-				    g_bytes_get_size(blob),
-				    self->fw_data_total_count, /* src */
-				    sizeof(self->public_key),
-				    error))
+		if (!fu_genesys_usbhub_device_read_flash(self,
+							 self->fw_key_addr[0],
+							 self->public_key,
+							 sizeof(self->public_key),
+							 NULL,
+							 error))
 			return FALSE;
 		guid = fwupd_guid_hash_data(self->public_key,
 					    sizeof(self->public_key),
 					    FWUPD_GUID_FLAG_NONE);
 		fu_device_add_instance_strup(device, "PUBKEY", guid);
+		if (!fu_device_build_instance_id(device, NULL, "USB", "VID", "PID", "PUBKEY", NULL))
+			return FALSE;
+	} else {
+		/* add specific product info */
+		if (self->running_bank == BANK_MASK_CODE) {
```

Comment:
you can move that outside the if statement. and have a `if (... != BANK_MASK_CODE)` with the process that was in the else statement.

This may share some code with the error IC type not supported code at line 759.